packages/keyring-api/src/superstruct.test.ts

```diff
@@ -1,72 +1,133 @@
+    { struct: nestedStruct, obj: { foo: { bar: 'hi' } }, expected: true },
+    {
+      struct: nestedStruct,
+      obj: { foo: { bar: undefined } },
+      expected: false,
+    },
+  ])(
+    'returns $expected for is($obj, <struct>)',
+    ({ struct, obj, expected }) => {
+      expect(is(obj, struct)).toBe(expected);
+    },
+  );
 
-    const structWithUndef = object({
-      foo: exactOptional(union([string(), literal(undefined)])),
+  const structWithUndef = object({
+    foo: exactOptional(union([string(), literal(undefined)])),
+  });
+
+  it.each([
+    { struct: structWithUndef, obj: {}, expected: true },
+    { struct: structWithUndef, obj: { foo: undefined }, expected: true },
+    { struct: structWithUndef, obj: { foo: 'hi' }, expected: true },
+    { struct: structWithUndef, obj: { bar: 'hi' }, expected: false },
+    { struct: structWithUndef, obj: { foo: 1 }, expected: false },
+  ])(
+    'returns $expected for is($obj, <struct>)',
+    ({ struct, obj, expected }) => {
+      expect(is(obj, struct)).toBe(expected);
+    },
+  );
+
+  it('should support refinements', () => {
+    const struct = object({
+      foo: exactOptional(max(number(), 0)),
     });
 
-    it.each([
-      { struct: structWithUndef, obj: {}, expected: true },
-      { struct: structWithUndef, obj: { foo: undefined }, expected: true },
-      { struct: structWithUndef, obj: { foo: 'hi' }, expected: true },
-      { struct: structWithUndef, obj: { bar: 'hi' }, expected: false },
-      { struct: structWithUndef, obj: { foo: 1 }, expected: false },
-    ])(
-      'returns $expected for is($obj, <struct>)',
-      ({ struct, obj, expected }) => {
-        expect(is(obj, struct)).toBe(expected);
-      },
+    expect(is({ foo: 0 }, struct)).toBe(true);
+    expect(is({ foo: -1 }, struct)).toBe(true);
+    expect(is({ foo: 1 }, struct)).toBe(false);
+  });
+});
+
+describe('selectiveUnion', () => {
+  const structA = object({
+    tag: literal('a'),
+    a: string(),
+  });
+
+  const structB = object({
+    tag: literal('b'),
+    b: coerce(number(), string(), (value) => parseFloat(value)),
+  });
+
+  const selector = (value: any): Struct<any, any> => {
+    return isPlainObject(value) && value.tag === 'a' ? structA : structB;
+  };
+
+  const struct = selectiveUnion(selector);
+
+  it('throws an error if the value does have the `tag` property', () => {
+    expect(() => assert({ other: 'c' }, struct)).toThrow(
+      'At path: tag -- Expected the literal `"b"`, but received: undefined',
     );
+  });
+
+  it('throws an error if the value has an invalid `tag`', () => {
+    expect(() => assert({ tag: 'c' }, struct)).toThrow(
+      'At path: tag -- Expected the literal `"b"`, but received: "c"',
+    );
+  });
 
-    it('should support refinements', () => {
-      const struct = object({
-        foo: exactOptional(max(number(), 0)),
-      });
+  it.each([
+    { obj: {}, expected: false },
+    { obj: { tag: 'c' }, expected: false },
```

Comment:
We are testing it just before? IDK if we need to keep it here too?

Reply:
Checked with @ccharly, we can keep both since the other test checks the error message.